new-log-viewer/src/contexts/NotificationContextProvider.tsx

```diff
@@ -0,0 +1,199 @@
+import React, {
+    createContext,
+    useCallback,
+    useRef,
+    useState,
+} from "react";
+
+import {
+    Alert,
+    Box,
+    IconButton,
+    Snackbar,
+    Stack,
+    Typography,
+} from "@mui/joy";
+
+import CloseIcon from "@mui/icons-material/Close";
+
+import {Nullable} from "../typings/common";
+import {LOG_LEVEL} from "../typings/logs";
+
+
+const AUTO_DISMISS_TIMEOUT_MILLIS = 5000;
+const DO_NOT_TIMEOUT_VALUE = null;
+
+/**
+ * Callback for posting a pop-up message with a title and level. The message is automatically
+ * dismissed after `timeoutMillis` if it is not {@link DO_NOT_TIMEOUT_VALUE}.
+ */
+type PostPopupCallback = (
+    level: LOG_LEVEL,
+    message: string,
+    title: string,
+    timeoutMillis: Nullable<number>
+) => void;
+
+/**
+ * Callback for posting a status message with level.
+ *
+ * When the level is less than or equal to `LOG_LEVEL.INFO`, the message is automatically dismissed
+ * after `AUTO_DISMISS_TIMEOUT_MILLIS`.
+ */
+type PostStatusCallback = (level: LOG_LEVEL, message: string, title?: string) => void;
+
+interface NotificationContextType {
+    statusMessage: string,
+
+    postPopup: PostPopupCallback,
+    postStatus: PostStatusCallback,
+}
+
+const NotificationContext = createContext<NotificationContextType>({} as NotificationContextType);
+
+interface NotificationContextProviderProps {
+    children: React.ReactNode;
+}
+
+/**
+ * Default values of the notification value object.
+ */
+const NOTIFICATION_DEFAULT: Readonly<NotificationContextType> = Object.freeze({
+    statusMessage: "",
+
+    postPopup: () => null,
+    postStatus: () => null,
+});
+
+interface PopupMessage {
+    level: LOG_LEVEL,
+    message: string,
+    title: string,
+    timeout: Nullable<ReturnType<typeof setTimeout>>
+}
+
+/**
+ * Provides notification management for the application. This provider must be at the outermost
+ * layer to ensure that both direct and indirect child components can publish and subscribe to
+ * notifications.
+ *
+ * @param props
+ * @param props.children
+ * @return
+ */
+const NotificationContextProvider = ({children}: NotificationContextProviderProps) => {
+    const [popupMessages, setPopupMessages] = useState<PopupMessage[]>([]);
+    const [statusMessage, setStatusMessage] = useState<string>(NOTIFICATION_DEFAULT.statusMessage);
+    const statusMsgTimeoutRef = useRef<Nullable<ReturnType<typeof setTimeout>>>(null);
+
+    const postPopup = useCallback((
+        level: LOG_LEVEL,
+        message: string,
+        title: string,
+        timeoutMillis: Nullable<number>
+    ) => {
+        const newMessage = {
+            level: level,
+            message: message,
+            title: "" === title ?
+                LOG_LEVEL[level] :
+                title,
+            timeout: DO_NOT_TIMEOUT_VALUE === timeoutMillis ?
+                null :
+                setTimeout(() => {
```

Comment:
> Create a popup component, and pass it a onClose() method - allowing it to delete itself similiar to the filter function underneath. You could also pass it the timeoutMillis. Then you could register the timeout with a ref in the popup component, and have it call onClose() after the timeout.

Note that we support creating multiple notification boxes (i.e. multiple pop-up messages can exist) at the same time. Does it mean that we will have to keep the components always mounted once created?

Reply:
Right, I think now there's no reason to store the return value of `setTimeout`. Before we support multiple pop-up messages, it was intended to cancel the timeout to avoid the new message getting erased due to the timeout callback.

Reply:
sg. implemented at e3463226ed602ceb43d720889d61e6d2ca47ec7a 

Note this implementation starts the timer only after `PopUpMessageBox` is mounted, which is even better imo.